### PR TITLE
Generic field access

### LINKED ISSUE #140: OverloadedLabels requires either manually writing instances or using TemplateHaskell
I really like `generic-lens` base approach. It was using existing library, and was so convenient.

The inconvinience is visible from the example

https://github.com/well-typed/optics/blob/f5c678166546c333e850579558fe9f9b7133c6c7/optics-core/src/Optics/Internal/Optic.hs#L191-L231

if optics wouldn't be scary topic to begin with!

### PR DESCRIPTION
**Pull request by arybczak**
While working on https://gitlab.haskell.org/ghc/ghc/-/merge_requests/2965 I extracted code from `generic-lens-lite` for a perf test and made it fast at compile time (and fast at runtime with the above MR).

Taking https://github.com/well-typed/optics/pull/359#issuecomment-703321123 into consideration I'm tentatively for removing `GeneralOpticLabel` (and the pluggable incoherence it introduces) and provide generic field access.

You may think I'm contradicting myself since I was against generic field access, but:

1) I didn't know that you can get `gfield` in under 100 lines of (modestly complicated) code with good error handling.
2) After #359 field access with labels will get confusing due to `setField` (depending on GHC version), plugabble generic based access and TH.
3) There is something to be said about the initial experience of a new user. If you can define a data type with a Generic instance and start using labels right away in a manner very similar to `RecordDotSyntax` without looking at TH and reading a guide in `Optics.Label`, it's a win.

Right now I'm thinking of
- Adding `gafield` to `Optics.AffineTraversal` for partial fields (similar effort as `gfield`).
- ~~Integrating both into LabelOptic so that you get a `Lens` or an `AffineTraversal` depending whether the field you want is partial or not.~~

So then with #359 we'd have a better story:
- Before GHC 9.2 you get *type-changing*, generic field access out of the box for convenience (and TH for ~~type changing updates and~~ speed)
- After GHC 9.2 you get field access for:
  * partial fields: type-preserving `Lens` that might fail with `HasField` or ~~`AffineTraversal`~~ type error if the type has Generic instance.
  * total fields: type-preserving `Lens` with `HasField` machinery ~~no matter if the type has a Generic instance~~ or type-changing `Lens` if the type has a Generic instance (and type-changing is needed).

~~Though it's not yet clear how expensive is checking if a field is total.~~

That would also fix #140.

Thoughts? @phadej @adamgundry

### CODE REVIEW

**Comment by adamgundry** [comment]
It'd be a shame to remove `GeneralLabelOptic` just after kcsongor/generic-lens#120 got in... on the other hand I do see that it would be nice to reduce the complexity by dropping the extension point. Would we simply drop any possibility of generic prisms (i.e. they would require TH/manual declarations)?

I like the idea of making partial fields safer and getting `AffineTraversal`s for them. It's a shame `HasField` doesn't have a nice way to enable this at the moment. But I'm a bit worried by the confusion that could result if partial fields use `HasField` to get a lens in the absence of a `Generic` instance, but a generic affine traversal if there is such an instance. And it does seem like checking partiality at each use site might get expensive for large datatypes?

A thought that someone else brought up recently: could we use `DerivingVia` to require explicit (but simple) instance declarations for fields but give the user more control?

I'll continue thinking about this and look at the code, but I'm away on holiday for the next couple of weeks, so may not have much availability for a while.

**Comment by arybczak** [comment]
> It'd be a shame to remove GeneralLabelOptic just after kcsongor/generic-lens#120 got in...

The timing is indeed a bit unfortunate, but it was laying there for a couple of months (and my other PRs are still not merged). I'd like `optics` to provide simple and fast (both at compile and run time) solution, `generic-lens` struggles a little in this department.

> Would we simply drop any possibility of generic prisms (i.e. they would require TH/manual declarations)?

I added prisms for constructors having up to 5 fields (should take care of 99% of cases).

> But I'm a bit worried by the confusion that could result if partial fields use HasField to get a lens in the absence of a Generic instance, but a generic affine traversal if there is such an instance. And it does seem like checking partiality at each use site might get expensive for large datatypes?

Right, it's a bit problematic. I think the best way would be to offer generic affine traversals as a function, but not integrated as labels to avoid the confusion after we integrate with `HasField`.

Also, it's better to not encourage usage of partial field selectors as they frankly suck.

**Comment by arybczak** [comment]
Ok, I created `Optics.Generic` with `GField`, `GAffineField` and `GConstructor` (also moved `GPlate` there) and modified `OpticLabel` to use `GField` and `GConstructor` if appropriate.

**Comment by arybczak** [comment]
I added support for type changing updates. Only one variant as they compile faster than type-preserving variants from `generic-lens` (not to mention type-changing ones) and have excellent type inference.

```haskell
λ> data X ph a b = X1 String a Int | X2 b deriving (Show, Generic)
λ> x1 :: Prism (X ph a b) _ _ _; x1 = #_X1

<interactive>:46:24: error:
    • Found type wildcard ‘_’ standing for ‘X b1 b2 b’
      Where: ‘b1’, ‘b2’, ‘b’ are rigid type variables bound by
               the inferred type of
                 x1 :: Prism
                         (X ph a b) (X b1 b2 b) ([Char], a, Int) ([Char], b2, Int)
               at <interactive>:46:1-39
      To use the inferred type, enable PartialTypeSignatures
    • In the second argument of ‘Prism’, namely ‘_’
      In the type ‘Prism (X ph a b) _ _ _’
      In the type signature: x1 :: Prism (X ph a b) _ _ _

<interactive>:46:26: error:
    • Found type wildcard ‘_’ standing for ‘([Char], a, Int)’
      Where: ‘a’ is a rigid type variable bound by
               the inferred type of
                 x1 :: Prism
                         (X ph a b) (X b1 b2 b) ([Char], a, Int) ([Char], b2, Int)
               at <interactive>:46:1-28
      To use the inferred type, enable PartialTypeSignatures
    • In the third argument of ‘Prism’, namely ‘_’
      In the type ‘Prism (X ph a b) _ _ _’
      In the type signature: x1 :: Prism (X ph a b) _ _ _

<interactive>:46:28: error:
    • Found type wildcard ‘_’ standing for ‘([Char], b2, Int)’
      Where: ‘b2’ is a rigid type variable bound by
               the inferred type of
                 x1 :: Prism
                         (X ph a b) (X b1 b2 b) ([Char], a, Int) ([Char], b2, Int)
               at <interactive>:46:31-39
      To use the inferred type, enable PartialTypeSignatures
    • In the fourth argument of ‘Prism’, namely ‘_’
      In the type ‘Prism (X ph a b) _ _ _’
      In the type signature: x1 :: Prism (X ph a b) _ _ _
```

```haskell
λ> x1 :: Prism _ (X ph a b) _ _; x1 = #_X1

<interactive>:47:13: error:
    • Found type wildcard ‘_’ standing for ‘X b1 b2 b’
      Where: ‘b1’, ‘b2’, ‘b’ are rigid type variables bound by
               the inferred type of
                 x1 :: Prism
                         (X b1 b2 b) (X ph a b) ([Char], b2, Int) ([Char], a, Int)
               at <interactive>:47:1-39
      To use the inferred type, enable PartialTypeSignatures
    • In the first argument of ‘Prism’, namely ‘_’
      In the type ‘Prism _ (X ph a b) _ _’
      In the type signature: x1 :: Prism _ (X ph a b) _ _

<interactive>:47:26: error:
    • Found type wildcard ‘_’ standing for ‘([Char], b2, Int)’
      Where: ‘b2’ is a rigid type variable bound by
               the inferred type of
                 x1 :: Prism
                         (X b1 b2 b) (X ph a b) ([Char], b2, Int) ([Char], a, Int)
               at <interactive>:47:31-39
      To use the inferred type, enable PartialTypeSignatures
    • In the third argument of ‘Prism’, namely ‘_’
      In the type ‘Prism _ (X ph a b) _ _’
      In the type signature: x1 :: Prism _ (X ph a b) _ _

<interactive>:47:28: error:
    • Found type wildcard ‘_’ standing for ‘([Char], a, Int)’
      Where: ‘a’ is a rigid type variable bound by
               the inferred type of
                 x1 :: Prism
                         (X b1 b2 b) (X ph a b) ([Char], b2, Int) ([Char], a, Int)
               at <interactive>:47:1-28
      To use the inferred type, enable PartialTypeSignatures
    • In the fourth argument of ‘Prism’, namely ‘_’
      In the type ‘Prism _ (X ph a b) _ _’
      In the type signature: x1 :: Prism _ (X ph a b) _ _
```

It can even change phantom type variables.

**Comment by arybczak** [comment]
Compile time benchmarks from https://github.com/arybczak/generic-lens/tree/perf-work (`cabal build generic-optics:perf`) with GHC 8.10.2.

`field'`,`field`, `position'` and `position` are from `generic-optics` library, `L.field` is from `generic-optics-lite` library (@phadej have a look, it's very inefficient).

1 constructor:

|variant   |memory |time  |
|----------|-------|------|
|gfield    |296M   |1.95s |
|field'    |342M   |2.91s |
|field     |329M   |3.37s |
|L.field   |1085M  |14.88s|
|          |       |      |
|gposition |304M   |2.05s |
|position' |1406M  |26.17s|
|position  |1022M  |66.01s|

2 constructors:

|variant   |memory |time  |
|----------|-------|------|
|gfield    |415M   |3.71s |
|field'    |544M   |5.97s |
|field     |592M   |6.69s |
|L.field   |1359M  |25.77s|
|          |       |      |
|gposition |406M   |3.90s |
|position' |2208M  |32.67s|
|position  |2993M  |73.46s|

3 constructors:

|variant   |memory |time  |
|----------|-------|------|
|gfield    |514M   |4.51s |
|field'    |714M   |7.75s |
|field     |827M   |9.07s |
|L.field   |1959M  |35.18s|
|          |       |      |
|gposition |507M   |4.81s |
|position' |2602M  |50.10s|
|position  |3848M  |108.9s|

**Comment by arybczak** [comment]
I've made `gfield` use `HasField` for reading and generics for setting so that reading is always fast even if generics don't optimize away.

**Comment by arybczak** [comment]
For the record, https://gitlab.haskell.org/ghc/ghc/-/merge_requests/2965 was merged so generics will optimize much better in GHC 9.2 in general.

**Comment by arybczak** [comment]
I've tested this on a real code base with about ~2k labels as optics (mostly used with `view`) (GHC 8.10.2).

|opt | labels  | memory [MB] | time [s] |
|----|---------|-------------|----------|
|0   | TH      | 1590        | 63       |
|0   | Generic | 2146        | 74       |
|1   | TH      | 5285        | 225      |
|1   | Generic | 5690        | 232      |

It's not nearly as bad as I thought it would be, it actually looks usable.

@adamgundry ping, please take a look when you have a moment.

**Review by adamgundry — COMMENTED** [review]
Apologies again for the delay in getting to this. It's really impressive! I can't say I follow all the details, but it would definitely improve the user experience to have support for generic derivation of optics out of the box. Moreover, https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3257 has run into the sand over compile-time performance, so I'm less optimistic that we will be able to switch to `HasField` for setting in the near term.

Assuming we go for this, we should flag up the revert needed in `generic-optics` and hopefully avoid a release of `generic-optics` that depends on `GeneralLabelOptic` being pluggable. I'm increasingly coming to feel that it is better to make it non-pluggable though and leave it as an implementation detail of `optics-core`.

I don't fully understand the `LiftCoverageCondition` trick, in particular the `TypeInferenceLoop` detection, and I remain slightly nervous that this may prove fragile. It's impressive you found a way to generate a reasonably nice error on reduction stack overflow, though I guess if the user sets `-freduction-depth=0` then they may just end up with a nonterminating compiler. Is it clear that the only looping case is in inference of local bindings subject to the monomorphism restriction? How worried should we be that this will break in future GHCs? I see your `COVERED` pragma proposal would address this, so perhaps we should see how that is received?

It feels like what we really want is a weaker notion of functional dependency: `LabelOptic` should require something subtly weaker than `name s b -> t`. This reminds me a bit of https://github.com/effectfully/sketches/tree/master/has-lens-done-right...

**Comment by arybczak** [comment]
> Assuming we go for this, we should flag up the revert needed in `generic-optics` and hopefully avoid a release of `generic-optics` that depends on `GeneralLabelOptic` being pluggable. I'm increasingly coming to feel that it is better to make it non-pluggable though and leave it as an implementation detail of `optics-core`.

I created https://github.com/kcsongor/generic-lens/issues/126 to keep track of this problem.

> I don't fully understand the `LiftCoverageCondition` trick, in particular the `TypeInferenceLoop` detection, and I remain slightly nervous that this may prove fragile.

`LiftCoverageCondition` includes the fundeps we want to lift and delegates to a class named `TypeInferenceLoop` with the same fundeps (and additional type parameters for the message) and the instance that lifts them with the circular trick. If GHC ends up looping on this constraint (as for how this might happen, see below), it'll report a `reduction stack overflow` error that contains the constraint it reached the limit on, namely the `TypeInfereceLoop` class with these somewhat helpful error messages.

There is nothing special about the `LiftCoverageCondition` there, it's just that instead of looping on the circular `GeneralLabelOptic` or `GField` constraint it'll loop on the `TypeInferenceLoop` constraint that when printed on `reduction stack overflow` gives you an idea about what's going on.

> It's impressive you found a way to generate a reasonably nice error on reduction stack overflow, though I guess if the user sets `-freduction-depth=0` then they may just end up with a nonterminating compiler.

Right, though it can be argued that this should never be done, so they brought that on themselves ;)

> Is it clear that the only looping case is in inference of local bindings subject to the monomorphism restriction?

It's not clear, but I tested with the code base that has 3.7k of labels in total and I didn't get this problem once. In fact, it doesn't even always happen in the case I mentioned, since e.g. the code base in question has this function:

```haskell
sortByEmail :: [User] -> [User]
sortByEmail = sortBy (\a b -> compare (f a) (f b)) where f = (^. #info % #email)
```

and it compiles fine.

> How worried should we be that this will break in future GHCs?

Not much. During constraint solving in GHC there are 3 kinds of constraints: Wanteds and Deriveds (and Givens, but they're not important here). Wanteds are cached, that's why the circular trick usually works, since when GHC loops once, it sees that the constraint was already solved and it stops. The problem is with Deriveds - they are not cached, so if GHC ends up with `LiftCoverageCondition` as a derived constraint, it'll loop. I'm not aware of all the circumstances in which this happens, but it seems to happen rarely. What is more, Richard Eisenberg is in the process of removal of Deriveds (he gave a talk about this on Haskell Exchange, the PR in progress is [here](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4149)).

With removal of Deriveds this should actually work reliably even without the proposal.


> I see your `COVERED` pragma proposal would address this, so perhaps we should see how that is received?

Right, though even if this isn't accepted, it looks like it's not going to be much of a problem.

> It feels like what we really want is a weaker notion of functional dependency: `LabelOptic` should require something subtly weaker than `name s b -> t`. This reminds me a bit of https://github.com/effectfully/sketches/tree/master/has-lens-done-right...

I'm afraid it's not that simple. If you look at the `GFieldImpl` class, it only has the `name s -> a` fundep. `name t -> b` fundep is not accepted because the path is generated from `Rep s`. We know that the path is the same for `Rep s` and `Rep t` from `HasShapeOf s t` and `HasShapeOf t s`, but GHC doesn't. So to get that fundep, we'd need to generate the path for `Rep s`, then for `Rep t` and them unify them. This slows things down considerably.

And as if this wasn't bad enough, this instance:

```haskell
instance
  ( GConstructorImpl name s t a b
  , _name ~ AppendSymbol "_" name
  ) => GenericOptic _name A_Prism s t a b where
  genericOptic = gconstructorImpl @name
```

cannot have any associated fundeps as even `name s -> a` is not accepted here because of the `AppendSymbol` trick.

**Comment by arybczak** [comment]
It needs to be noted that the `LabelOptic` for pluggable generic optics on master exhibits the same "loopiness" problem because of https://github.com/well-typed/optics/blob/master/optics-core/src/Optics/Internal/Optic.hs#L210 (except without a reasonable error message of `reduction stack overflow`).

File: optics-core/src/Optics/Label.hs [hunk 1]
```diff
@@ -454,5 +472,103 @@ import Optics.Internal.Optic
...
+-- | Otherwise report an error.
+instance {-# INCOHERENT #-}
+  ( TypeError
+    ('Text "No instance for LabelOptic " ':<>: 'ShowType name
+     ':<>: 'Text " " ':<>: QuoteType k
+     ':<>: 'Text " " ':<>: QuoteType s
+     ':<>: 'Text " " ':<>: QuoteType t
+     ':<>: 'Text " " ':<>: QuoteType a
+     ':<>: 'Text " " ':<>: QuoteType b
+     ':$$: 'Text "Perhaps you forgot to define it or misspelled its name?")
```
[the remark this answers is not in the document]
Comment (arybczak) [reply optics-core/src/Optics/Label.hs]: Done.

**Comment by arybczak** [comment]
I'm inclined to have it merged since (as I mentioned above) the funky fundep situation is no worse than with pluggable `GeneralOpticLabel` and this PR takes a huge step forward by making generic optics usable in real-world scenarios. Any objections?

**Comment by arybczak** [comment]
I'll do a few finishing touches on the documentation and merge during the weekend unless anyone shouts.

**Comment by adamgundry** [comment]
Thank you for continuing to push on this. I'm unfortunately lacking time to go through in great detail, but the general plan sounds okay to me.

**Comment by phadej** [comment]
One question: Would it be possible to **never** resolve `LabelOptic` via `GHC.Generics`? If the functionality can be disabled with a cabal flag, I don't care about the details.

**Comment by arybczak** [comment]
@phadej We could add a flag, though that might be a bit confusing. What's the motivation?

**Comment by phadej** [comment]
**I don't (infinitely) trust GHC optimizer.** I don't understand how manually defined `LabelOptic` instances and generics would interact, and I'd prefer to get compilation error because of no instance then a *silent* fallback to generics.

**Comment by arybczak** [comment]
> I don't (infinitely) trust GHC optimizer

There are two components to the whole story:

1. Resolution of `LabelOptic` instances.
2. Optimization of generic field access if the generic `LabelOptic` instance is picked.

What you hint at is the first point, whereas the optimizer is involved only in the second one. Instance resolution is predictable.

For the `LabelOptic` instance to be reliably picked by GHC it has to have `a`, `b` and `k` free in the instance head and constrained in the instance context. That follows from functional dependencies of `LabelOptic` and does not depend on the optimization level. TH generated instances will always be preferred to the generic instance, so we're covered here.

If we had fundeps sans the coverage condition (https://github.com/ghc-proposals/ghc-proposals/pull/374), the fundeps on `LabelOptic` could be changed to `name s -> k t a b, name t -> k s a b` and then we wouldn't have to be precise about free variables in the instance head, but sadly this is not the case yet.

To elaborate further, if the generic `LabelOptic` is picked, the resulting lens will *always* be equivalent in performance to the regular field accessor in terms of reading because it uses `HasField` for getting the field. It uses generics only for setting, so then if the data type is large enough (>10 fields for a pure product) and they don't optimize away it's going to be a couple of times slower than hand-written/TH generated one.

Prisms are similar to the setting case for fields - generics don't optimize away for pure sums with >8 constructors.

This is solved in GHC 9.2 because of https://gitlab.haskell.org/ghc/ghc/-/merge_requests/2965 - there are regression tests for a pure product with 100 fields (though it can be arbitrarily large really) and pure sum with 16 constructors. Though then hopefully we'll also get `SetField` so it won't be needed for fields anymore.

---------------

We can alter the overlappable instance with the cabal flag, but I personally am not a fan of having a flag change API in such a subtle way. I spent a lot of time fine-tuning generic optics precisely so that you don't have to worry about performance (see https://github.com/well-typed/optics/pull/361#issuecomment-722328934 for a real-world testing).

@adamgundry any thoughts?

**Comment by phadej** [comment]
> It uses generics only for setting, so then if the data type is large enough (>10 fields for a pure product) and they don't optimize away it's going to be a couple of times slower than hand-written/TH generated one.

I would prefer that to be an error rather than a (hard to spot in large) slowdown, if I forget to TH generate an instance.

**Comment by arybczak** [comment]
> I would prefer that to be an error rather than a (hard to spot in large) slowdown, if I forget to TH generate an instance.

Right, I see your point. I'm just unsure how to solve it in a satisfactory way.

I thought this PR will allow libraries to use the labels with `optics-core` :(

**Comment by arybczak** [comment]
@phadej does https://github.com/well-typed/optics/pull/361/commits/65194d5935c5d53251539c3167113a738a9826db (Require GenericLabelOptics instance for generic labels) satisfy you?

**Comment by phadej** [comment]
@arybczak yes, that is great for me. But indeed, I'm quite sure there are people who would prefer having generic labels for everything on by default. I don't see a way to satisfy everyone (thus why I'm fine with flag for myself).

**Comment by arybczak** [comment]
Yeah. That covers the case of libraries wanting to use generic optics with their data types and applications wanting to use generic optics in general (as if a specific library is missing the instances, orphans can be provided), but it doesn't work for libraries wanting to use generic optics with data types from other libraries not knowing about `optics` :(

Ideally I wanted this to create a sort of compatibility layer for sane field handling from now on until we get `SetField` - as then the overlappable instance in `optics` would just use `SetField` for setting instead of generics and no changes would be necessary to code that uses `optics`.

But I do see the problem of using generics silently if one forgets to define `LabelOptic` instances (either by hand or via TH)...

At least this is better than the cabal flag, since with a flag no library can use generic labels.

I'll let it marinate for another week or so, maybe someone will have an enlightenment how to solve it gracefully.

**Comment by adamgundry** [comment]
> At least this is better than the cabal flag, since with a flag no library can use generic labels.

What about having a manual flag biased towards allowing generic labels:
 * By default, labels use generics silently. Libraries are implicitly allowed to assume this will be the case.
 * If the flag is set, the generic instance is replaced with one that uses `TypeError`.

That way users who really care about performance can set the flag at build time and guarantee the absence of generic instances. Of course they might hit the problem that some library they depend on uses a generic instance from another library they depend on, and the build will fail. But if they did explicitly ask for that...

> We can alter the overlappable instance with the cabal flag, but I personally am not a fan of having a flag change API in such a subtle way.

I generally agree that flags shouldn't change APIs, but in this case we are really just saying we want to build in "hard mode" as an additional check on the build. And if that is explicitly requested by the person building the package, it seems okay to me. 

Really we want some way to (optionally) request a warning when a certain instance is used, which is occasionally requested as a GHC feature.

**Comment by arybczak** [comment]
@adamgundry I was thinking about a similar thing - manual flag, enabled by default and explicit note that turning it off is generally unsupported.

That would render the flag unusable once/if libraries rely on it enough. On the other hand, the `GenericLabelOptics` approach would most likely mean for labels to never gain enough traction :slightly_smiling_face: at least before `SetField`.

Choices...

**Comment by arybczak** [comment]
I think I have a middle ground solution.

I added `explicit-generic-labels` flag to optics-core (disabled by default). When disabled, only the `Generic` instance is needed for the overlappable `LabelOptic` instance to fire. When enabled, `GenericLabelOptics` instance is needed in addition.

This gives more flexibility to application authors as they can enable this flag and patch/send PRs to affected libraries to simply derive `GenericLabelOptics` (or include specific `LabelOptic` instances) for offending data types.

**Comment by arybczak** [comment]
This looks satisfactory to me, I'll merge (this time for sure :stuck_out_tongue: ) during the weekend.

**Review by adamgundry — APPROVED** [review]
Thanks, I'm happy with this approach. I've made some suggestions regarding documentation.

I was initially worried that it is unclear under what circumstances it would make sense for libraries to provide `GenericLabelOptics` instances, and hence wondering whether the class is useful in addition to the flag. But I think it is fine to say they shouldn't normally bother, yet keep the class as an override mechanism so that an application developer can choose to build with `-fexplicit-generic-labels` and potentially define (possibly orphan) `GenericLabelOptics` instances.

**Comment by arybczak** [comment]
> But I think it is fine to say they shouldn't normally bother

Libraries shouldn't bother unless they themselves not use generic labels with their data types, then it makes sense for them to derive these instances so that the library works with `explicit-generic-labels` (I added this to the doc).

File: optics-core/src/Optics/Label.hs [hunk 2]
```diff
@@ -1,3 +1,10 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeInType #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
[the remark this answers is not in the document]
Comment (arybczak) [reply optics-core/src/Optics/Label.hs:7]: Indeed, I didn't see a better way.

**Comment by arybczak** [comment]
@adamgundry I applied your suggestions, I think we're good to merge this :tada:

**Pull request merged**